tianshou/policy/modelbase/psrl.py

```diff
@@ -0,0 +1,212 @@
+import numpy as np
+from typing import Any, Dict, Union, Optional
+
+from tianshou.data import Batch
+from tianshou.policy import BasePolicy
+
+
+class PSRLModel(object):
+    """Implementation of Posterior Sampling Reinforcement Learning Model.
+
+    :param np.ndarray trans_count_prior: dirichlet prior (alphas), with shape
+        (n_state, n_action, n_state).
+    :param np.ndarray rew_mean_prior: means of the normal priors of rewards,
+        with shape (n_state, n_action).
+    :param np.ndarray rew_std_prior: standard deviations of the normal priors
+        of rewards, with shape (n_state, n_action).
+    :param float discount_factor: in [0, 1].
+    :param float epsilon: for precision control in value iteration.
+    """
+
+    def __init__(
+        self,
+        trans_count_prior: np.ndarray,
+        rew_mean_prior: np.ndarray,
+        rew_std_prior: np.ndarray,
+        discount_factor: float,
+        epsilon: float,
+    ) -> None:
+        self.trans_count = trans_count_prior
+        self.n_state, self.n_action = rew_mean_prior.shape
+        self.rew_mean = rew_mean_prior
+        self.rew_std = rew_std_prior
+        self.discount_factor = discount_factor
+        self.rew_count = np.full(rew_mean_prior.shape, epsilon)  # no weight
+        self.eps = epsilon
+        self.policy: np.ndarray
+        self.value = np.zeros(self.n_state)
+        self.updated = False
+
+    def observe(
+        self,
+        trans_count: np.ndarray,
+        rew_sum: np.ndarray,
+        rew_count: np.ndarray
+    ) -> None:
+        """Add data into memory pool.
+
+        For rewards, we have a normal prior at first. After we observed a
+        reward for a given state-action pair, we use the mean value of our
+        observations instead of the prior mean as the posterior mean. The
+        standard deviations are in inverse proportion to the number of the
+        corresponding observations.
+
+        :param np.ndarray trans_count: the number of observations, with shape
+            (n_state, n_action, n_state).
+        :param np.ndarray rew_sum: total rewards, with shape
+            (n_state, n_action).
+        :param np.ndarray rew_count: the number of rewards, with shape
+            (n_state, n_action).
+        """
+        self.updated = False
+        self.trans_count += trans_count
+        sum_count = self.rew_count + rew_count
+        self.rew_mean = (self.rew_mean * self.rew_count + rew_sum) / sum_count
+        self.rew_std *= self.rew_count / sum_count
```

Comment:
This is not true. This is only the approximation. To correctly calculate std, we need to maintain rew_square_sum and let `std = sqrt(rew_square_sum / rew_count - rew_mean ** 2)`.